Integrate build-cache cleanup with global cache configuration

bigdaz opened the pull request:
Fixes #28505 

### Context
Reliable and predictable cleanup of Gradle User Home is important to users attempting to manage disk usage, both locally and on CI. This PR allows the cleanup of the local build cache to be configured via the [regular cache configuration mechanisms](https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home:configure_cache_cleanup).

bigdaz commented on subprojects/core-api/src/main/java/org/gradle/caching/local/DirectoryBuildCache.java:
```diff
@@ -28,7 +28,7 @@
 public abstract class DirectoryBuildCache extends AbstractBuildCache {
     private Object directory;
-    private int removeUnusedEntriesAfterDays = 7;
+    private int removeUnusedEntriesAfterDays = -1;
```
I suppose this is a breaking change, if anyone is checking the default value programmatically. The effective default value is the same, since this is taken from `CacheConfiguration.createdResources()`, which has a 7 day default.

asodja replied:
❓ What about if we add some internal property that will return if `removeUnusedEntriesAfterDays` was set or not?

bigdaz replied:
This change was reverted: the default value is again `7`.

asodja commented on platforms/core-execution/build-cache-local/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheServiceFactory.java:
```diff
@@ -81,11 +85,16 @@ public BuildCacheService createBuildCacheService(DirectoryBuildCache configurati
             config("removeUnusedEntriesAfter", removeUnusedEntriesAfterDays + " days");
 
+        // Use the days configured for the local build cache, or the global 'createdResources' config if not specified.
+        Supplier<Long> removeUnusedEntriesOlderThan = removeUnusedEntriesAfterDays < 1
```
🤔 I see that when we deprecated `org.gradle.cache.cleanup` [user home definition overrides project definiton](https://github.com/gradle/gradle/blob/master/platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc#deprecated-orggradlecachecleanup-property). But here it's the other way around.

I wonder if we should the same strategy here?

asodja replied:
Oh, that property was under user home.

bigdaz replied:
Unfortunately, to make the new `buildCache.setRemoveUnusedEntriesAfterDays` take precedence over the (now deprecated) `DirectoryBuildCache` method would require significant complexity in the former.
With this PR, if a user calls `DirectoryBuildCache.setRemoveUnusedEntriesAfterDays()` with a non-default value, then they will get a deprecation warning but the value will be honoured. If they don't call that deprecated method, then the value will be obtained via the new API.

asodja commented:
It would be nice if we also document this in https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home:configure_cache_cleanup

and here:
https://docs.gradle.org/current/userguide/build_cache.html#sec:build_cache_configure_local

Maybe as a follow-up PR?

lkasso reviewed (approved):
Docs (user manual + javadocs) LGTM.

bigdaz commented:
@asodja This PR is ready for another round of review:
- Added `CacheConfigurations.buildCache`, which can be configured just like `createdResources` (and other Gradle User Home cleanup).
- Deprecated `DirectoryBuildCache.removeUnusedEntriesAfterDays`. Users will be nagged when this value is set.
- Reverted the previous default for `DirectoryBuildCache.removeUnusedEntriesAfterDays`. Behaviour is now:
  - If the user has changed this from the default, the changed value will be used.
  - If the user has not set this, or has set it to the default value (7), then the value for `CacheConfigurations.buildCache` will be used. This value also defaults to 7 days.
  - For ease of implementation, the `DirectoryBuildCache` setting takes precedence. This will only impact users who are setting _both_ the now-deprecated `DirectoryBuildCache` and the brand new `CacheConfigurations.buildCache` values.

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"900d4306ab5d3648dc0cad777c5a1eaecbcb3b81","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
I've triggered the following builds for you. Click [here](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.names=gitCommitId&amp;search.tags=CI&amp;search.values=900d4306ab5d3648dc0cad777c5a1eaecbcb3b81) to see all build failures.
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=80206663)

bigdaz commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"e90ddaf6e2dfdadebbd15f23900b8da8a642d7f5","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
I've triggered the following builds for you. Click [here](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.names=gitCommitId&amp;search.tags=CI&amp;search.values=e90ddaf6e2dfdadebbd15f23900b8da8a642d7f5) to see all build failures.
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=80213184)

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"e90ddaf6e2dfdadebbd15f23900b8da8a642d7f5","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
WARN: This is a feature pull request but no changes to [release note](https://github.com/gradle/gradle/blob/master/platforms/documentation/docs/src/docs/release/notes.md) found.

asodja reviewed (approved):
LGTM

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"e90ddaf6e2dfdadebbd15f23900b8da8a642d7f5","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
WARN: This is a feature pull request but no changes to [release note](https://github.com/gradle/gradle/blob/master/platforms/documentation/docs/src/docs/release/notes.md) found.

bot-gradle commented:
[The merge queue build](https://builds.gradle.org/viewQueued.html?itemId=80223980) has started. Click [here](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.names=gitCommitId&amp;search.tags=CI&amp;search.values=baf05dfb6ca6dc301c358655e7bf8a1bebe3464c) to see all failures if any.

bot-gradle commented:
[The merge queue build](https://builds.gradle.org/viewQueued.html?itemId=80237093) has started. Click [here](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.names=gitCommitId&amp;search.tags=CI&amp;search.values=22ebab694b7f880d62bf64941ed3f15816670fe3) to see all failures if any.
